onnxruntime/core/framework/provider_bridge_ort.cc

```diff
@@ -504,12 +504,26 @@ struct ProviderHostImpl : ProviderHost {
 
 struct ProviderLibrary {
   ProviderLibrary(const char* filename) {
-
     std::string full_path = Env::Default().GetRuntimePath() + std::string(filename);
     Env::Default().LoadDynamicLibrary(full_path, &handle_);
     if (!handle_)
       return;
 
+#ifdef _WIN32
+    {
+      // We crash when unloading DNNL on windows due to OpenMP (As there are threads running code inside the openmp runtime DLL if
+      // OMP_WAIT_POLICY is set to ACTIVE). To avoid this, we pin the OpenMP DLL so that it unloads as late as possible.
+      HMODULE handle{};
+#ifdef _DEBUG
+      constexpr const char* dll_name = "vcomp140d.dll";
+#else
+      constexpr const char* dll_name = "vcomp140.dll";
```

Comment:
this assumes we build with omp library supplied by VC++
in the past, we could build ORT/dnnl with with omp library (libiomp5md.dll) supplied by intel (which is binary compat). it is part of mklml package. 
i thought we added this capability in our cmake scripts, but maybe it was never added officially.
just FYI. maybe don't need to change this PR. but maybe should document it as a caveat somewhere.

Reply:
Should this be done only if we've built with openmp?

Reply:
if you build dnnl, by default it uses openmp as the threading runtime. (but it is possible to configure dnnl to build with different threading library)
so, this gets a little messy. 
technically,
you can build ORT with openmp and dnnl with openmp ,
or you could build ORT without openmp and dnnl with openmp
or you can build ORT without openmp and dnnl without openmp.
or you could build dnnl with a intel's openmp.

Reply:
Yeah, I agree it makes a lot of assumptions but without it we crash on exit when OpenMP starts running. It would be interesting if the Intel one didn't have this flaw (it's safe to try to pin this DLL every time, even if it's not even loaded).

Reply:
@pranavsharma , if we've built with openmp and the dnnl also uses the same openmp runtime, then it's ok, it won't crash even without this change.

Reply:
@pranavsharma I think it's the opposite. You only need this if ORT is not using OMP (if it was you would already have a dependency at the ORT level on the OMP library) and the library we're unloading is using OMP.

It would be nice if we could detect at build time whether an EP library is using OMP and force ORT to also use OMP if that is the case so that we don't have two different threadpool implementations being used. Not sure that's possible longer term though if we're going to build these EP libraries separately to ORT.

Reply:
Yeah, I mistyped. I meant ensuring if the specific EP DLL was built with/using openmp.
Regarding 2 different threadpool impls, I remember Stuart was talking to Intel to make the threadpool configurable. Not sure if that converged. @jywu-msft might know.

Reply:
see https://oneapi-src.github.io/oneDNN/dev_guide_threadpool.html
(we are using an older version of DNNL though - renamed to oneDNN)